server/conn.go

```diff
@@ -715,6 +763,34 @@ func (cc *clientConn) openSessionAndDoAuth(authData []byte) error {
 	return nil
 }
 
+func (cc *clientConn) checkAuthPlugin(ctx context.Context, authPlugin *string) ([]byte, error) {
+	// Open a context unless this was done before.
+	if cc.ctx == nil {
+		cc.openSession()
+	}
+
+	host, _, err := cc.PeerHost("")
+	if err != nil {
+		return nil, err
+	}
+
+	userplugin, err := cc.ctx.AuthPluginForUser(&auth.UserIdentity{Username: cc.user, Hostname: host})
+	if len(userplugin) == 0 {
+		*authPlugin = mysql.AuthNativePassword
+		return nil, nil
+	} else if (cc.authPlugin != userplugin) || (cc.authPlugin != *authPlugin) {
+		authData, err := cc.authSwitchRequest(ctx, userplugin)
+		if err != nil {
+			return nil, err
+		}
+		*authPlugin = userplugin
+		return authData, nil
+	} else {
+		*authPlugin = userplugin
+	}
```

Comment:
The linter should warn that this else branch should be removed, since the previous if / else if both return.